### PR TITLE
[# 151900718] Write a blog post on GraphQL authorization

### PR DESCRIPTION
**Pull request by etripier**
Write a blog post on GraphQL authorization

[Tracker #151900718](https://www.pivotaltracker.com/story/show/151900718)

### CODE REVIEW

File: source/_posts/2017-10-09-guarding-an-open-door-authorizing-your-graphql-fields.markdown [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+  # Returns the initial value for `memo`
+  def initial_value(query)
+  end
+
+  # This is like the `reduce` callback.
+  # The return value is passed to the next call as `memo`
+  def call(memo, visit_type, irep_node)
+  end
+
+  # Called when we're done the whole visit.
```
Comment (runongirlrunon) [inline source/_posts/2017-10-09-guarding-an-open-door-authorizing-your-graphql-fields.markdown]: I think you accidentally a word.

Comment (etripier) [reply source/_posts/2017-10-09-guarding-an-open-door-authorizing-your-graphql-fields.markdown]: Heh, this is a straight copy from the GraphQL Ruby docs: http://graphql-ruby.org/queries/analysis.html. Good catch, though!

File: source/_posts/2017-10-09-guarding-an-open-door-authorizing-your-graphql-fields.markdown [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+def call(memo, visit_type, irep_node)
+  irep_node.definition # nil
+end
+```
+
+The first time we hit this method, our `irep_node` will be the `query` node (one of three [root types](http://graphql.org/learn/execution/#root-fields-resolvers) in our schema: `query`, `mutation`, and `subscription`). The second time, we'll get the `delayedJobs` node:
+
+```ruby
+def call(memo, visit_type, irep_node)
+  irep_node.definition.name # "delayedJobs"
```
[the remark this answers is not in the document]
Comment (etripier) [reply source/_posts/2017-10-09-guarding-an-open-door-authorizing-your-graphql-fields.markdown]: In the line above, we're inspecting the `query` root type, which doesn't have a `definition` - so we can't call `#name` on it.

File: source/_posts/2017-10-09-guarding-an-open-door-authorizing-your-graphql-fields.markdown [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+end
+```
+
+Checking for the right permissions goes something like this:
+
+```ruby
+def call(memo, visit_type, irep_node)
+  current_user = irep_node.query.context[:current_user]
+  requested_node_name = irep_node.definition.name.to_sym
+  ability = Ability.new(current_user) # Using CanCan
```
[the remark this answers is not in the document]
Comment (etripier) [reply source/_posts/2017-10-09-guarding-an-open-door-authorizing-your-graphql-fields.markdown]: Yep! I'll fix that.

File: source/_posts/2017-10-09-guarding-an-open-door-authorizing-your-graphql-fields.markdown [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+## Runtime Authorization
+Let's go back to the third-grade class in the lobby. We're not worried if they're playing with the gold bar on display; in fact, it's encouraged (what else is there to do at the Mint?). We'd be a little more concerned if we found a third-grader in the vaults, though.
+
+In that case, we can't forbid everyone from handling our gold. But we can be very clear about *which* ingots can be touched.
+
+Let's take a look at a different query:
+
+```
+query {
+  user(id: 123) {
```
Comment (beldaGR) [inline source/_posts/2017-10-09-guarding-an-open-door-authorizing-your-graphql-fields.markdown]: Should we use a model that doesn't exist in our db (maybe related to a mint)? Or is user generic enough? Just double-checking.

Comment (etripier) [reply source/_posts/2017-10-09-guarding-an-open-door-authorizing-your-graphql-fields.markdown]: My assumption is that `User` is generic enough that it doesn't reveal too much about our schema, and that it's relevant enough to be a good example, but I'm definitely open to changing these examples.

Comment (beldaGR) [reply source/_posts/2017-10-09-guarding-an-open-door-authorizing-your-graphql-fields.markdown]: Ok, I'm fine with this example unless someone else pipes up.

**Comment by etripier** [comment]
Thanks for the reviews! I hope I've fixed all the outstanding issues.

**Pull request merged**